tests/suite/graceful_recovery_test.go

```diff
@@ -0,0 +1,342 @@
+)
+
+const (
+	nginxContainerName = "nginx"
+	ngfContainerName   = "nginx-gateway"
+)
+
+// Since checkContainerLogsForErrors may experience interference from previous tests (as explained in the function
+// documentation), this test is recommended to be run separate from other nfr tests.
+var _ = Describe("Graceful Recovery test", Ordered, Label("nfr", "graceful-recovery"), func() {
+	files := []string{
+		"graceful-recovery/cafe.yaml",
+		"graceful-recovery/cafe-secret.yaml",
+		"graceful-recovery/gateway.yaml",
+		"graceful-recovery/cafe-routes.yaml",
+	}
+
+	ns := &core.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "graceful-recovery",
+		},
+	}
+
+	teaURL := "https://cafe.example.com/tea"
+	coffeeURL := "http://cafe.example.com/coffee"
+
+	var ngfPodName string
+
+	BeforeAll(func() {
+		podNames, err := framework.GetReadyNGFPodNames(k8sClient, ngfNamespace, releaseName, timeoutConfig.GetTimeout)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(podNames).ToNot(BeEmpty())
+
+		ngfPodName = podNames[0]
+	})
+
+	BeforeEach(func() {
+		Expect(resourceManager.Apply([]client.Object{ns})).To(Succeed())
+		Expect(resourceManager.ApplyFromFiles(files, ns.Name)).To(Succeed())
+		Expect(resourceManager.WaitForAppsToBeReady(ns.Name)).To(Succeed())
+
+		Eventually(
+			func() bool {
+				return checkForWorkingTraffic(teaURL, coffeeURL)
+			}).
+			WithTimeout(timeoutConfig.RequestTimeout).
+			WithPolling(500 * time.Millisecond).
+			Should(BeTrue())
+	})
+
+	AfterAll(func() {
+		Expect(resourceManager.DeleteFromFiles(files, ns.Name)).To(Succeed())
+		Expect(resourceManager.Delete([]client.Object{ns})).To(Succeed())
+	})
+
+	It("recovers when NGF container is restarted", func() {
+		runRecoveryTest(teaURL, coffeeURL, ngfPodName, ngfContainerName, files, ns)
+	})
+
+	It("recovers when nginx container is restarted", func() {
+		// FIXME(bjee19) remove Skip() when https://github.com/nginxinc/nginx-gateway-fabric/issues/1108 is completed.
+		Skip("Test currently fails due to this issue: https://github.com/nginxinc/nginx-gateway-fabric/issues/1108")
+		runRecoveryTest(teaURL, coffeeURL, ngfPodName, nginxContainerName, files, ns)
+	})
+})
+
+func runRecoveryTest(teaURL, coffeeURL, ngfPodName, containerName string, files []string, ns *core.Namespace) {
+	var (
+		err       error
+		leaseName string
+	)
+
+	if containerName != nginxContainerName {
+		// Since we have already deployed resources and ran resourceManager.WaitForAppsToBeReady(ns.Name) earlier,
+		// we know that the applications are ready at this point. This could only be the case if NGF has written
+		// statuses, which could only be the case if NGF has the leader lease. Since there is only one instance
+		// of NGF in this test, we can be certain that this is the correct leaseholder name.
+		leaseName, err = getLeaderElectionLeaseHolderName()
+		Expect(err).ToNot(HaveOccurred())
+	}
+
+	restartContainer(ngfPodName, containerName)
+
+	checkContainerLogsForErrors(ngfPodName)
```

Comment:
Let's check the logs after other things. This way, if errors occur as the results of actions below, we will catch them.

Reply:
Actually removed the check here and moved it after all the other actions since the logs collected should encompass all of it. Let me know if this was a poor idea.

Reply:
Update:

On some runs (very small amount) I see in the nginx logs an `[error]` caused by a HTTP request getting a 502 bad gateway response. This is expected as we allow for some failing traffic, which is why we poll to until the traffic passes. However, this gets caught in our `checkContainerLogsForErrors` and the test fails because of a single `[error]` in the nginx logs. 

```
  2024/05/01 21:18:29 [error] 73#73: *21 connect() failed (111: Connection refused) while connecting to upstream, client: 10.128.0.69, server: cafe.example.com, request: "GET /tea HTTP/1.1", upstream: "http://10.120.2.139:8080/tea", host: "cafe.example.com"
  10.128.0.69 - - [01/May/2024:21:18:29 +0000] "GET /tea HTTP/1.1" 502 157 "-" "Go-http-client/1.1"

```

After some consideration I can see a couple possible avenues to solve this
- We can remove the check in logs for `[error]` meaning we will only check on `[emerg], [crit], [alert]` 
- We can try to add back &sinceSeconds on our log checks and figure out some kind of "hack"
- I can try to figure out some way to allow for 502 errors in the logs.

@pleshakov what do you think?

Reply:
> I can try to figure out some way to allow for 502 errors in the logs.

Is it possible to filter out those?

for example, we can split the logs by line
and then check it is has 
(1) [emerg], [crit], [alert]
(2) [error]. if it has error, it must not be "connect() failed (111: Connection refused)"

Reply:
Added a fix, @pleshakov let me know your thoughts.

Tested that after this line was printed in the logs
```
2024/05/02 22:09:54 [error] 75#75: *27 connect() failed (111: Connection refused) while connecting to upstream, client: 10.128.0.69, server: cafe.example.com, request: "GET /tea HTTP/1.1", upstream: "http://10.120.2.149:8080/tea", host: "cafe.example.com"
10.128.0.69 - - [02/May/2024:22:09:54 +0000] "GET /tea HTTP/1.1" 502 157 "-" "Go-http-client/1.1"

```

The test still passed.

Reply:
If this is an expected error and the only one we know of right now, I think the fix looks good.